### DISCUSSION TITLE
Lambda function URLs and Compression

### DISCUSSION THREAD

**Discussion by genifycom** [aws/aws-lambda-dotnet#1168] [Q&A] [answered]
I have started using Lambda function URLs and have a question.
I have an object that I would like to return as Brotli compressed to the caller.
1. I serialize my_object to a JSON string
2. I Brotli compress the JSON string
3. I Base64 the compressed string
4. I return the Base64 string and content type of br

Here is the code:
```
            var json = JsonSerializer.Serialize(my_object);
             var compressedBytes = await Compressor.BrotliCompressBytesAsync(System.Text.Encoding.UTF8.GetBytes(json), cancel);
             var payload = Convert.ToBase64String(compressedBytes);
             return new APIGatewayProxyResponse
                {
                    StatusCode = (int)HttpStatusCode.OK,
                    Body = payload,
                    Headers = new Dictionary<string, string> { { "Content-Type", "application/json" }, { "Content-Encoding", "br" } }
                };
```
The result is that Postman tell me "Error: Decompression failed".

I am assuming that something happens in the Lambda function URL like "Content type conversions in API Gateway".

Is there a content header that tells the Lambda function URL to leave the body alone or is there a better way to do this?

Why am I doing this? Because of the Lambda payload size constraint of 6MB.

**Comment by ashishdhingra** [comment, marked as answer]
Hi @genifycom,

Thanks for posting the question. The following articles might be helpful:
- [Content type conversions in API Gateway](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-payload-encodings-workflow.html) (There is a note at the bottom `To have the binary payload passed through without modification, you must enable the passthrough behaviors on the API.`)
- [Integration passthrough behaviors](https://docs.aws.amazon.com/apigateway/latest/developerguide/integration-passthrough-behaviors.html)

Please review.

Thanks,
Ashish

**Reply by genifycom** [reply]
Thanks for responding Ashish. I am aware of this with an API Gateway. However, this is using a Lambda Function URL and in the Console I am not able to find the API Gateway associated with this. There are no triggers or obviously attached items.

Can you give me more information for the Lambda Function URL case please?

**Reply by ashishdhingra** [reply]
@genifycom Lambda Function URLs and API Gateway are all together different concepts. Few helpful articles:
- [Announcing AWS Lambda Function URLs: Built-in HTTPS Endpoints for Single-Function Microservices](https://aws.amazon.com/blogs/aws/announcing-aws-lambda-function-urls-built-in-https-endpoints-for-single-function-microservices/)
- [Exploring AWS Lambda Function URL using .NET Container - Part 1](https://dev.to/aws-builders/exploring-aws-lambda-function-url-using-net-container-eg8)
- [AWS Lambda Function URLs vs. Amazon API Gateway](https://www.serverlessguru.com/blog/aws-lambda-function-urls-vs-amazon-api-gateway)

Also as **Request and response payloads** section at [Invoking Lambda function URLs](https://docs.aws.amazon.com/lambda/latest/dg/urls-invocation.html) mentions, the request and response event formats follow the same schema as the [Amazon API Gateway payload format version 2.0](https://docs.aws.amazon.com/apigateway/latest/developerguide/http-api-develop-integrations-lambda.html#http-api-develop-integrations-lambda.proxy-format). Based on your example above, even though your original object was serialized as JSON, you are converting into binary format by compressing it into bytes and then base64 encoding it. This essentially means your response is a Base64 encoded compressed binary format, however, the content type you are setting in response is `application/json`. I'm unsure how Lambda function URLs work behind the scenes, since your return content type is `application/json`, but your response body is base64 encoded compressed bytes, the Lambda function URL deserialization process to JSON fails.

Since Lambda function URLs point to simple Lambda functions, ASP.NET Core runtime in serverless scenario with API Gateway is out of picture, and HTTP response compression as available in ASP.NET Core framework is unavailable. I'm unsure on how this could be achieved for Lambda function URL scenario.

**Comment by genifycom** [comment]
Thanks for that. BTW, I did try application/octet-stream but that didn't seem to help. I will have to investigate further, or attach the function to an API gateway.
